File: Python/16987_계란으로계란치기.py
import sys

def breakingEgg(N, s, egg, n_broken, cur_broken):
	if s == N: # 가장 최근에 든 계란이 가장 오른쪽에 위치한 계란일 경우
		return n_broken # 종료
	
	if egg[s][0] <= 0: # 현재 선택된 계란이 이미 깨진 계란인 경우
		n_broken = breakingEgg(N,s+1,egg,n_broken,cur_broken) # 한 칸 옆 계란 선택
		return n_broken
	
	for i in range(N): # 깨지지 않은 계란 찾기
		if i == s or egg[i][0] <= 0: # 현재 들고 있는 계란이거나, 깨진 계란 선택했을 경우
			continue # 건너 뛰기

		# 깨지지 않은 다른 계란 선택했을 때		
		egg[i][0] -= egg[s][1]
		egg[s][0] -= egg[i][1] # 서로 충돌했기 때문에 내구성 감소(상대방의 무게만큼)
		
		sub = 0 # 깨진 계란의 수 세기
		if egg[i][0] <= 0: # 선택한 계란 깨졌는지 확인
			sub += 1
		if egg[s][0] <= 0: # 들고 있는 계란 깨졌는지 확인
			sub += 1
		cur_broken += sub # 현재 충돌로 인해 깨진 계란만큼 깨진 계란의 수 증가
		
		if n_broken < cur_broken:
			n_broken = cur_broken # 깨진 계란의 수의 최댓값 확인하여 갱신(max 함수 쓰니까 시간초과..ㅠ)
		n_broken = breakingEgg(N,s+1,egg,n_broken,cur_broken) # 현재 계란 내려 놓고 오른쪽 옆에 있는 계란 선택
		
		egg[i][0] += egg[s][1]
		egg[s][0] += egg[i][1]
		cur_broken -= sub # 백트래킹
		
	return n_broken
	

if __name__ == '__main__':
	N = int(sys.stdin.readline())
	egg = [list(map(int,sys.stdin.readline().split())) for _ in range(N)]
	
	n_broken = breakingEgg(N,0,egg,0,0)

	print(n_broken)

File: C++/16987_계란으로계란치기.cpp
#include <iostream>
#include <algorithm>
using namespace std;

struct Egg{
	int s, w;
}; // 계란의 내구성, 무게를 담을 struct

void breakingEgg(int N, int s, Egg *egg, int &n_broken, int cur_broken);

int main() {
	int N;
	cin >> N;
	
	Egg *egg = new Egg[N];
	for(int i = 0; i < N; i++){
		cin >> egg[i].s >> egg[i].w;
	}
	
	int n_broken = 0; // 깨진 계란의 최대값 담을 변수
	breakingEgg(N,0,egg,n_broken,0);
	cout << n_broken;
	
	return 0;
}

void breakingEgg(int N, int s, Egg *egg, int &n_broken, int cur_broken){
	if(s == N){ // 가장 최근에 든 계란이 가장 오른쪽에 위치한 계란일 경우
		return; // 더 이상 진행하지 않고 종료
	}
	
	if(egg[s].s <= 0){ // 현재 들고 있는 계란이 깨진 계란일 경우
		breakingEgg(N,s+1,egg,n_broken,cur_broken); // 다음 계란으로 넘어가기
		return;
	}
	
	// 현재 들고 있는 계란이 깨지지 않은 계란일 경우
	for(int i = 0; i < N; i++){ // 계란 선택
		if(i == s || egg[i].s <= 0){ // 선택한 계란이 들고 있는 계란과 같은 계란이거나, 깨진 계란일 경우
			continue; // 아무 것도 수행하지 않고 다음 계란 선택하러 이동..ㅎ
		}
		
		egg[i].s -= egg[s].w;
		egg[s].s -= egg[i].w; // 충돌하면 상대 계란의 무게만큼 내구성 감소
		
		int sub = 0; // 현재 충돌로 인해 몇 개의 계란이 깨졌는지
		if(egg[i].s <= 0){ // 선택한 계란이 깨졌는지 확인
			sub++;
		}
		if(egg[s].s <= 0){ // 들고 있던 계란이 깨졌는지 확인
			sub++;
		}
		cur_broken += sub; // 현재 충돌로 인해 깨진 계란만큼 깨진 계란의 수 증가
		n_broken = max(n_broken,cur_broken); // 깨진 계란의 최대 개수 갱신
		
		breakingEgg(N,s+1,egg,n_broken,cur_broken); // 현재 들고 있는 계란 놓고 다음 계란 들기
		
		egg[i].s += egg[s].w;
		egg[s].s += egg[i].w;
		cur_broken -= sub; // 백트래킹
	}
}
